src/routers/alpha-router/gas-models/v3/v3-heuristic-gas-model.ts

```diff
@@ -259,6 +269,81 @@ export class V3HeuristicGasModelFactory extends IOnChainGasModelFactory {
         throw err;
       }
 
+      /*
+        We want to get the gas cost in terms of the quote token.
+        Sometimes, the quote token / native token pair is very illiquid and the price is incorrect.
+        We also have access to the input token / native token pair, which is more likely to be liquid.
+          - if it's not, then it is an illiquid direct swap trade
+        We can get the gas cost in terms of the input token, 
+          using the inputTokenNativePool and quoting for the amount of ETH required for gas
+        Then, we can use the execution price to get the respective amounts of quote tokens
+        Then, we can take the minimum of the initial quoted gas cost in terms of quote token
+          and our calculated one.
+      */
+      try {
+        // get current execution price (inputToken / quoteToken)
+        const executionPrice = new Price(
+          routeWithValidQuote.amount.currency,
+          routeWithValidQuote.quote.currency,
+          routeWithValidQuote.amount.quotient,
+          routeWithValidQuote.quote.quotient
+        );
+        // Highest liquidity pool for the non quote token / ETH
+        // A pool with the non quote token / ETH should not be required and errors should be handled separately
+        if (nativeInputPool) {
+          const inputIsToken0 =
+            nativeInputPool.token0.address == nativeCurrency.address;
+          // ratio of input / native
+          const nativeInputTokenPrice = inputIsToken0
+            ? nativeInputPool.token0Price
+            : nativeInputPool.token1Price;
+
+          const gasCostInTermsOfInputToken = nativeInputTokenPrice.quote(
+            totalGasCostNativeCurrency
+          ) as CurrencyAmount;
+
+          // Convert gasCostInTermsOfInputToken to quote token using execution price
+          const syntheticGasCostInTermsOfQuoteToken = executionPrice.quote(
+            gasCostInTermsOfInputToken
+          );
+
+          // If the cost in quote tokens over the execution price is lessThan the cost found using the quoteToken/native pool
+          // then we use our synthetic gas cost in tokens
+          if (
+            syntheticGasCostInTermsOfQuoteToken.lessThan(
+              gasCostInTermsOfQuoteToken.asFraction
+            )
+          ) {
```

Comment:
Thinking about this more, I'm not sure that if the synthetic is smaller then it is *always* better. Like it would be possible for there to be a borked amount/native pool with a price such that it always returns a smaller + less accurate value. It just happens that the USDP pool is broken in a way where the quoting ETH returns a huge value.

Still though, I think its better to prefer the smaller value since big values lead to crazy results because the adjusted quotes go negative. Where as with smaller values we kind of just fallback to ignoring gas when picking the best route- which feels a lot better.

Does that makes sense? If so i think we should call it out in the comment

Reply:
Yeah I agree. If the synthetic price calculated is extremely small due to error, we will basically just return the quote w/o gas consideration, and at least the user receives a quote - we use the gasEstimate in gwei anyways when submitting so shouldnt be a huge issue? Maybe it would hit slippage, i'm not sure as I haven't looked at that code in a while.

Reply:
Additionally, something interesting that I've noticed is that the vast majority of the integ tests and cli quotes that I've run result in a smaller syntheticGasCost... it's smaller by a small amount, and I'm not quite sure why.